_posts/2025-05-09-sorting-for-fast-selective-queries.md

```diff
@@ -0,0 +1,660 @@
+        floor(rowid / 122880) AS rowgroup_id,
+        ' || dq_concat(column_list, ',') || '
+    ), rowgroup_id_counts AS (
+    FROM by_rowgroup_id
+    SELECT
+        case ' ||
+        nq_concat(list_transform(column_list, (i) -> ' when grouping('||dq(i)||') = 0 then alias('||dq(i)||') '),' ')
+            || ' end AS column_name,
+        coalesce(*columns(* exclude (rowgroup_id, total_row groups))) AS column_value,
+        first(total_row groups) AS total_row groups,
+        count(distinct rowgroup_id) AS rowgroup_id_count
+    GROUP BY
+        GROUPING SETS ( ' || nq_concat(list_transform(dq_list(column_list), (j) -> '('||j||')'), ', ') ||' )
+    )
+    FROM rowgroup_id_counts
+    SELECT
+        '||sq(table_name)||' AS table_name,
+        *
+    ORDER BY
+        column_name
+    ')
+);
+
+-- This is an optional function that can summarize the data
+-- as an alternative to boxplot charts
+CREATE OR REPLACE FUNCTION summarize_rowgroup_counts(table_name, column_list) AS TABLE (
+    FROM rowgroup_counts(table_name, column_list)
+    SELECT
+        table_name,
+        column_name,
+        total_row groups,
+        min(rowgroup_id_count) AS min_cluster_depth,
+        avg(rowgroup_id_count) AS avg_cluster_depth,
+        max(rowgroup_id_count) AS max_cluster_depth,
+        map([0.1, 0.25, 0.5, 0.75, 0.9], quantile_cont(rowgroup_id_count, [0.1, 0.25, 0.5, 0.75, 0.9]))::JSON AS quantiles,
+        histogram(rowgroup_id_count, [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 32, 64, 128, 256])::JSON AS histograms,
+    GROUP BY ALL
+    ORDER BY ALL
+);
+```
+
+</details>
+
+We can then call the `rowgroup_counts` function on any table and any columns!
+
+```sql
+FROM rowgroup_counts('flights_hilbert', ['origin', 'dest']);
+```
+
+| table_name      | column_name | column_value | total_row groups | rowgroup_id_count |
+| :-------------- | :---------- | :----------- | --------------: | ----------------: |
+| flights_hilbert | dest        | PSG          |             238 |                 2 |
+| flights_hilbert | dest        | ESC          |             238 |                 2 |
+| flights_hilbert | dest        | YUM          |             238 |                 2 |
+| flights_hilbert | dest        | TWF          |             238 |                 2 |
+| flights_hilbert | dest        | TUL          |             238 |                 7 |
+| ...             | ...         | ...          |             ... |               ... |
+
+The `rowgroup_id_count` column is a measurement of how many distinct row groups that a specific column value is present in, so it is an indicator of how much work DuckDB would need to do to pull all data associated with that value.
+
+> This calculation uses the [pseudo-column `rowid`]({% link docs/preview/sql/statements/select.md %}#row-ids), and it requires data to have been inserted in a single batch to be perfectly accurate.
+> It is directionally correct for data inserted in batches.
```

Comment:
What does "directionally correct" mean?

Reply:
Directionally correct is another way to say "approximate". I'm happy to change it to say approximate if that is clearer!

Reply:
Made that change!